File: nic/upgrade_manager/export/upgcsdk/upgrade_resp_reactor.cc
// {C} Copyright 2018 Pensando Systems Inc. All rights reserved.

#include <stdio.h>
#include <iostream>

#include "upgrade_resp_reactor.hpp"
#include "nic/upgrade_manager/utils/upgrade_log.hpp"

namespace upgrade {

using namespace std;
extern UpgCtx ctx;

string UpgRespReact::GetRespStr(delphi::objects::UpgRespPtr resp) {
    switch (resp->upgrespval()) {
        case UpgRespPass:
            return ("Upgrade Successful");
        case UpgRespFail:
            return ("Upgrade Failed");
        case UpgRespAbort:
            return ("Upgrade Aborted");
        default:
            return ("");
    }
}

delphi::error UpgRespReact::DeleteUpgReqSpec(void) {
    delphi::objects::UpgReqPtr req = delphi::objects::UpgReq::FindObject(sdk_);
    sdk_->DeleteObject(req);
    return delphi::error::OK();
}

void UpgRespReact::InvokeAgentHandler(delphi::objects::UpgRespPtr resp) {
    vector<string> errStrList;
    switch (resp->upgrespval()) {
        case UpgRespPass:
            upgAgentHandler_->UpgSuccessful();
            break;
        case UpgRespFail:
            for (int i=0; i<resp->upgrespfailstr_size(); i++) {
                errStrList.push_back(resp->upgrespfailstr(i));
            }
            upgAgentHandler_->UpgFailed(errStrList);
            break;
        case UpgRespAbort:
            for (int i=0; i<resp->upgrespfailstr_size(); i++) {
                errStrList.push_back(resp->upgrespfailstr(i));
            }
            upgAgentHandler_->UpgAborted(errStrList);
            break;
        case UpgRespUpgPossible:
            if (!resp->upgrespfailstr().empty()) {
                for (int i=0; i<resp->upgrespfailstr_size(); i++) {
                    errStrList.push_back(resp->upgrespfailstr(i));
                }
                upgAgentHandler_->UpgNotPossible(ctx, errStrList);
            } else {
                upgAgentHandler_->UpgPossible(ctx);
            }
            break;
        default:
            break;
    }
}

delphi::error UpgRespReact::OnUpgRespCreate(delphi::objects::UpgRespPtr resp) {
    UPG_LOG_DEBUG("UpgRespHdlr::OnUpgRespCreate called with status {}", GetRespStr(resp));
    if (DeleteUpgReqSpec() == delphi::error::OK()) {
        UPG_LOG_DEBUG("Upgrade Req Object deleted for next request");
    }
    InvokeAgentHandler(resp);
    return delphi::error::OK();
}

delphi::error UpgRespReact::OnUpgRespVal(delphi::objects::UpgRespPtr
resp) {
    if (GetRespStr(resp) != "")
        UPG_LOG_DEBUG("UpgRespHdlr::OnUpgRespVal called with status: {}", GetRespStr(resp));
    if (DeleteUpgReqSpec() == delphi::error::OK()) {
        UPG_LOG_DEBUG("Upgrade Req Object deleted for next request");
    }
    InvokeAgentHandler(resp);
    return delphi::error::OK();
}

delphi::objects::UpgRespPtr UpgRespReact::FindUpgRespSpec(void) {
    return delphi::objects::UpgResp::FindObject(sdk_);
}

void UpgRespReact::OnMountComplete(void) {
    delphi::objects::UpgRespPtr resp = delphi::objects::UpgResp::FindObject(sdk_);
    if (resp == NULL) {
        return;
    }
    OnUpgRespCreate(resp);
}

};

File: nic/upgrade_manager/export/upggosdk/upgrade_resp_reactor.go
package upggosdk

import (
	"github.com/pensando/sw/nic/delphi/gosdk/client_api"
	"github.com/pensando/sw/nic/delphi/proto/delphi"
	"github.com/pensando/sw/nic/upgrade_manager/export/upggosdk/nic/upgrade_manager/upgrade"
	"github.com/pensando/sw/venice/utils/log"
)

type upgrespctx struct {
	agentHdlrs AgentHandlers
	sdkClient  clientApi.Client
}

var upgRespCtx upgrespctx

func (ctx *upgrespctx) invokeAgentHandler(obj *upgrade.UpgResp) {
	var errStrList []string
	switch obj.GetUpgRespVal() {
	case upgrade.UpgRespType_UpgRespPass:
		log.Infof("upgrade success!!")
		ctx.agentHdlrs.UpgSuccessful()
	case upgrade.UpgRespType_UpgRespFail:
		log.Infof("upgrade failed!!")
		for _, str := range obj.GetUpgRespFailStr() {
			errStrList = append(errStrList, str)
		}
		ctx.agentHdlrs.UpgFailed(&errStrList)
	case upgrade.UpgRespType_UpgRespAbort:
		log.Infof("upgrade aborted!!")
		for _, str := range obj.GetUpgRespFailStr() {
			errStrList = append(errStrList, str)
		}
		ctx.agentHdlrs.UpgAborted(&errStrList)
	case upgrade.UpgRespType_UpgRespUpgPossible:
		log.Infof("%d length", len(obj.GetUpgRespFailStr()))
		if len(obj.GetUpgRespFailStr()) == 0 {
			log.Infof("upgrade possible")
			ctx.agentHdlrs.UpgPossible(&upgCtx)
		} else {
			for _, str := range obj.GetUpgRespFailStr() {
				errStrList = append(errStrList, str)
			}
			log.Infof("upgrade not possible")
			ctx.agentHdlrs.UpgNotPossible(&upgCtx, &errStrList)
		}
	}
}

func (ctx *upgrespctx) deleteUpgReqSpec() {
	upgReq := upgrade.GetUpgReq(ctx.sdkClient)
	if upgReq == nil {
		log.Infof("upgReq not found")
		return
	}
	ctx.sdkClient.DeleteObject(upgReq)
	log.Infof("Upgrade Req Object deleted for next request")
}

func (ctx *upgrespctx) OnUpgRespCreate(obj *upgrade.UpgResp) {
	log.Infof("OnUpgRespCreate called %d", obj.GetUpgRespVal())
	ctx.deleteUpgReqSpec()
	ctx.invokeAgentHandler(obj)
}

func (ctx *upgrespctx) OnUpgRespUpdate(old, obj *upgrade.UpgResp) {
	log.Infof("OnUpgRespUpdate called %d", obj.GetUpgRespVal())
	ctx.deleteUpgReqSpec()
	ctx.invokeAgentHandler(obj)
}

func (ctx *upgrespctx) OnUpgRespDelete(obj *upgrade.UpgResp) {
	log.Infof("OnUpgRespDelete called %d", obj.GetUpgRespVal())
}

//upgRespInit init resp subtree coming from upgrade manager
func upgRespInit(client clientApi.Client, hdlrs AgentHandlers) {
	log.Infof("upgRespInit called")
	ctx := &upgrespctx{
		agentHdlrs: hdlrs,
		sdkClient:  client,
	}
	upgrade.UpgRespMount(client, delphi.MountMode_ReadMode)
	upgrade.UpgRespWatch(client, ctx)
	client.WatchMount(ctx)
	upgRespCtx = *ctx
}

func (ctx *upgrespctx) OnMountComplete() {
	log.Infof("OnMountComplete got called to restore UpgResp")
	upgresp := upgrade.GetUpgResp(ctx.sdkClient)
	if upgresp == nil {
		log.Infof("OnMountComplete could not find UpgResp")
		return
	}
	ctx.OnUpgRespCreate(upgresp)
}
